optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,422 @@
+                    distribution, distribution_list
+                )
+            )
+
+        assert low < high
+
+        return low, high, q
+
+    def _transform_to_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
+            else:
+                distribution_list = [
+                    distributions.UniformDistribution.__name__,
+                    distributions.LogUniformDistribution.__name__,
+                    distributions.DiscreteUniformDistribution.__name__,
+                    distributions.IntUniformDistribution.__name__,
+                    distributions.IntLogUniformDistribution.__name__,
+                    distributions.CategoricalDistribution.__name__,
+                ]
+                raise NotImplementedError(
+                    "The distribution {} is not implemented. "
+                    "The parameter distribution should be one of the {}".format(
+                        distribution, distribution_list
+                    )
+                )
+        return transformed
+
+    def _transform_from_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.exp(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                q = self._q[param_name]
+                low = self._low[param_name]
+                high = self._high[param_name]
+                samples = (np.round((samples - low) / q - 0.5) + 0.5) * q + low
+                transformed[param_name] = np.clip(samples, low, high)
```

Comment:
These `low` and `high` are transformed `low` and `hight` in `_calculate_parzen_bounds`. We should use the original `distribution.low` and `distribution.high`.
```suggestion
                transformed[param_name] = np.clip(samples, distribution.low, distribution.high)
```

Reply:
Thank you for pointing this out! It was a critical bug.